packages/svelte2tsx/src/nodes/ComponentEvents.ts

```diff
@@ -0,0 +1,129 @@
+import ts from 'typescript';
+import { EventHandler } from './event-handler';
+import { getVariableAtTopLevel } from '../utils/tsAst';
+
+export abstract class ComponentEvents {
+    protected events = new Map<string, { type: string; doc?: string }>();
+
+    getAll(): { name: string; type?: string; doc?: string }[] {
+        const entries: { name: string; type: string; doc?: string }[] = [];
+
+        const iterableEntries = this.events.entries();
+        for (const entry of iterableEntries) {
+            entries.push({ name: entry[0], ...entry[1] });
+        }
+
+        return entries;
+    }
+
+    abstract toDefString(): string;
+}
+
+export class ComponentEventsFromInterface extends ComponentEvents {
+    constructor(node: ts.InterfaceDeclaration) {
+        super();
+        this.events = this.extractEvents(node);
+    }
+
+    toDefString() {
+        return '{} as unknown as ComponentEvents';
+    }
+
+    private extractEvents(node: ts.InterfaceDeclaration) {
+        const map = new Map<string, { type: string; doc?: string }>();
+
+        node.members.filter(ts.isPropertySignature).forEach((member) => {
+            map.set(this.getName(member.name), {
+                type: member.type?.getText() || 'Event',
+                doc: this.getDoc(node, member),
+            });
+        });
+
+        return map;
+    }
+
+    private getName(prop: ts.PropertyName) {
+        if (ts.isIdentifier(prop) || ts.isStringLiteral(prop)) {
+            return prop.text;
+        }
+
+        if (ts.isComputedPropertyName(prop)) {
+            if (ts.isIdentifier(prop.expression)) {
+                const identifierName = prop.expression.text;
+                const identifierValue = this.getIdentifierValue(prop, identifierName);
+                if (!identifierValue) {
+                    this.throwError(prop);
+                }
+                return identifierValue;
+            }
+        }
+
+        this.throwError(prop);
+    }
+
+    private getIdentifierValue(prop: ts.ComputedPropertyName, identifierName: string) {
+        const variable = getVariableAtTopLevel(prop.getSourceFile(), identifierName);
+        if (variable && ts.isStringLiteral(variable.initializer)) {
+            return variable.initializer.text;
+        }
+    }
+
+    private throwError(prop: ts.PropertyName) {
+        const error: any = new Error(
+            'The ComponentEvents interface can only have properties of type ' +
+                'Identifier, StringLiteral or ComputedPropertyName. ' +
+                'In case of ComputedPropertyName, ' +
+                'it must be a const declared within the component and initialized with a string.',
+        );
+        error.start = toLineColumn(prop.getStart());
+        error.end = toLineColumn(prop.getEnd());
+        throw error;
```

Comment:
This throws if the interface definition does not match our expextations. I hope this makes it clear to the user what is possible and what is not.